include/l2/upper_mac_packet_builder.hpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2024 Transit Live Mapping Solutions
+ * All rights reserved.
+ *
+ * Authors:
+ *   Marenz Schmidl
+ */
+
+#pragma once
+
+#include "l2/slot.hpp"
+#include "l2/upper_mac_packet.hpp"
+#include "utils/bit_vector.hpp"
+#include <cstddef>
+#include <optional>
+#include <stdexcept>
+#include <vector>
+
+/// datastructure to hold the packets that are processed in the upper MAC
+struct UpperMacPackets {
+    /// the C-plane signalling packets
+    std::vector<UpperMacCPlaneSignallingPacket> c_plane_signalling_packets_;
+    /// the U-plane signalling packets
+    std::vector<UpperMacUPlaneSignallingPacket> u_plane_signalling_packet_;
+    /// an optional U-plane traffic packet
+    std::optional<UpperMacUPlaneTrafficPacket> u_plane_traffic_packet_;
+    /// an optional broadcast packet
+    std::optional<UpperMacBroadcastPacket> broadcast_packet_;
+
+    /// Merge two extracted UpperMacPackets into one. Thsi function is used to merge the extracted packets from two
+    /// slots.
+    /// \param other the UpperMacPackets which should be merged into the current one
+    auto merge(UpperMacPackets&& other) -> void {
+        std::move(other.c_plane_signalling_packets_.begin(), other.c_plane_signalling_packets_.end(),
+                  std::back_inserter(c_plane_signalling_packets_));
+        std::move(other.u_plane_signalling_packet_.begin(), other.u_plane_signalling_packet_.end(),
+                  std::back_inserter(u_plane_signalling_packet_));
+
+        if (u_plane_traffic_packet_ && other.u_plane_traffic_packet_) {
+            throw std::runtime_error("Trying to merge two packets both with a UpperMacUPlaneTrafficPacket");
```

Comment:
I think in the long run we probably shouldn't use hard exceptions. 

Is it possible to create a malformed packet which would take the software down?

Reply:
As far as I read the standard this should not happen. We will discard Slots that trigger a runtime_error during parsing. The expectation is that these slots will be logged. See https://github.com/tlm-solutions/tetra-decoder/issues/15 and https://github.com/tlm-solutions/tetra-decoder/issues/16